File: Fourier/fourier.c
#include <math.h>

typedef struct
{
    float re; // really
    float im; // imaginary
} complex, *pcomplex;

complex complexAdd(complex a, complex b)
{
    complex ret;
    ret.re = a.re + b.re;
    ret.im = a.im + b.im;
    return ret;
}

complex complexMult(complex a, complex b)
{
    complex ret;
    ret.re = a.re * b.re - a.im * b.im;
    ret.im = a.im * b.re + a.re * b.im;
    return ret;
}

void DFT(complex x[], complex X[], int N)
{
    int k, n;
    complex Wnk;
    for (k = 0; k\< N; k++)
    {
        X[k].re = 0;
        X[k].im = 0;
        for (n = 0; n\< N; n++)
        {
            Wnk.re = (float)cos(2 * PI * k * n / N);
            Wnk.im = (float)-sin(2 * PI * k * n / N);
            X[k] = complexAdd(X[k], complexMult(x[n], Wnk));
        }
    }
}

int main()
{
    complex samples[10], _out[10];
    int i;
    for (i = 0; i\< 10; i++)
    {
        samples[i].re = i;
        samples[i].im = 0;
    }
    DFT(samples, _out, 10);
    for (i = 0; i\< 10; i++)
    {
        printf("(%f,%f)\n", _out[i].re, _out[i].im);
    }
}

File: Fourier/fourier.js
class Complex {
    constructor(re, im) {
        this.re = re;
        this.im = im;
    }
}

/**
 * 
 * @param {Complex} complex1 
 * @param {Complex} complex2 
 */
function complexAdd(complex1, complex2) {
    let re, im;
    re = complex1.re + complex2.re;
    im = complex1.im + complex2.im;
    return new Complex(re, im);
}

function complexMulti(complex1, complex2) {
    let re, im;
    re = complex1.re * complex2.re - complex1.im * complex2.im;
    im = complex1.re * complex2.im + complex1.im * complex2.re;
    return new Complex(re, im);
}

function DFT(x, N) {
    let X = new Array(N);
    let k, n;
    let wnk = new Complex(0, 0);
    for (k = 0; k < N; k++) {
        X[k] = new Complex(0, 0);
        for (n = 0; n < N; n++) {
            wnk.re = Math.cos(2 * Math.PI * k * n / N);
            wnk.im = -Math.sin(2 * Math.PI * k * n / N);
            X[k] = complexAdd(X[k], complexMulti(x[n], wnk));
        }
    }
    return X;
}


function IDFT(X, N) {
    let x = new Array(N);
    let k, n;
    let ejw = new Complex(0, 0);
    for (k = 0; k < N; k++) {
        x[k] = new Complex(0, 0);
        for (n = 0; n < N; n++) {
            ejw.re = Math.cos(2 * Math.PI * k * n / N);
            ejw.im = Math.sin(2 * Math.PI * k * n / N);
            x[k] = complexAdd(x[k], complexMulti(X[n], ejw));
            x[k].re /= N;
            x[k].im /= N;
        }
    }
    return x;
}
let x = [];
let n = 10;
for (let i = 1; i <= n; i++) {
    // let complex = new Complex(i * Math.PI / 180, 0);
    let complex = new Complex(i, 0);
    x.push(complex);
}
let X = DFT(x, n);
let ix = IDFT(x, n);
console.log(X);
console.log(ix);